src/libertem/executor/pipelined.py

```diff
@@ -0,0 +1,857 @@
+                                assert header_type not in (
+                                    "RUN_TASK", "SCATTER", "RUN_FUNCTION",
+                                    "CLEANUP", "SHUTDOWN", "WARMUP",
+                                )
+                                if header_type == "END_TASK":
+                                    break
+            elif header_type == "SCATTER":
+                # FIXME: array data could be transferred and stored in SHM instead
+                key = header["key"]
+                assert key not in work_mem
+                value = header["value"]
+                work_mem[key] = value
+                continue
+            elif header_type == "RUN_FUNCTION":
+                with attach_to_parent(header["span_context"]):
+                    worker_run_function(header, queues, worker_idx)
+            elif header_type == "CLEANUP":
+                key = header["key"]
+                assert key in work_mem
+                del work_mem[key]
+                continue
+            elif header_type == "SHUTDOWN":
+                with attach_to_parent(header["span_context"]):
+                    with tracer.start_as_current_span("SHUTDOWN") as span:
+                        queues.request.close()
+                        queues.response.close()
+                    break
+            elif header_type == "WARMUP":
+                with attach_to_parent(header["span_context"]):
+                    with tracer.start_as_current_span("WARMUP"):
+                        import libertem.udf.base  # NOQA
+                        import libertem.api  # NOQA
+                        import libertem.preload  # NOQA
+                        with env.enter():
+                            pass
+            else:
+                queues.response.put({
+                    "type": "ERROR",
+                    "error": f"unknown message {header}",
+                    "worker_id": worker_idx,
+                })
+                continue
+
+
+DeviceT = Tuple[int, Union[Literal['CUDA'], Literal['CPU']]]
+
+
+def _setup_device(spec: WorkerSpec, pin: bool):
+    """
+    Set up this worker for its given task - either CPU or GPU comptation,
+    and maybe pin CPU workers to a given CPU core.
+    """
+    if spec["device_kind"].lower() == "cpu":
+        if hasattr(os, 'sched_setaffinity') and pin:
+            os.sched_setaffinity(0, [spec["device_id"]])
+        set_use_cpu(spec["device_id"])
+    elif spec["device_kind"].lower() == "cuda":
+        set_use_cuda(spec["device_id"])
+
+
+def _drain_and_close(queues: WorkerQueues):
+    """
+    Drain request queue, close request and response queue,
+    Called in case the worker encounters an unhandled exception.
+    """
+    while True:
+        try:
+            with queues.request.get(block=False):
+                continue
+        except WorkerQueueEmpty:
+            break
+    # close and join:
+    queues.request.close()
```

Comment:
You are right. I think this might also be an issue: I think we must not drain the response queue here, on the worker, as that would mean responses can get lost before they reach the main process (i.e. written by the worker process and directly drained out again)

This might need a change in the `WorkerQueue` interface, I'll need to think about this, but it should be possible to do backwards-compatible.

Reply:
> I think we must not drain the response queue here, on the worker, as that would mean responses can get lost before they reach the main process
Good spot, I didn't think about that!

> but it should be possible to do backwards-compatible.
Agreed.

Reply:
`drain` parameter is now available for `WorkerQueue.close`